### ISSUE TITLE
Make the single-line use of [<Literal>] consistent

### ISSUE THREAD

**Issue by baronfel** [fsharp/fslang-suggestions#409]
# Make the single-line use of [<Literal>] consistent [7090326]

**Submitted by Don Syme on 2/13/2015 12:00:00 AM**  
**6 votes on UserVoice prior to migration**  

See https://github.com/Microsoft/visualfsharp/issues/229
Some declaration constructs allow attributes on the same line as the keyword for the declaration, e.g.

```fsharp
[<Measure>] type kg
[<Measure>] type m
```

However this doesn't work for "let" bindings, e.g.
![img](https://cloud.githubusercontent.com/assets/445888/6168930/cd4fc538-b2c9-11e4-8961-6efa3f0e4080.png)
Right now the consistent way to do things is always to use multiple lines, e.g.
![img2](https://cloud.githubusercontent.com/assets/445888/6168957/07eb7688-b2ca-11e4-887e-97e8de7c90c9.png)
The suggestion is to make this consistent by allowing single-line use everywhere, if that is technically possible.

## Response

**by fslang-admin on 2/5/2016 12:00:00 AM**

> Closing as this idea (listed by me) as it has little support

**[Original UserVoice Submission](https://fslang.uservoice.com/forums/245727-f-language/suggestions/7090326)**
[Archived Uservoice Comments](../tree/master/archive/suggestion-7090326-make-the-single-line-use-of-[-literal-]-consisten.md#comments)

**Comment by Happypig375** [comment]
Please resurrect this, I want this so bad 😢

**Comment by cartermp** [comment]
Thanks for the ping. I'd like to relitigate this, since the closure was based on lack of vote count, which I don't really see as a valid reason to decline something.

@dsyme thoughts?

**Comment by baronfel** [comment]
Edited the original submission for legibility/inline image expansion.

**Comment by FunctionalFirst** [comment]
This is already possible (or am I misunderstanding the suggestion?).
```f#
let [<Literal>] Foo = "foo"
```

**Comment by cartermp** [comment]
@FunctionalFirst Correct, but it's a bit unexpected that it has to come after the `let`. If you do it as described in the issue, you get an offside warning:

![image](https://user-images.githubusercontent.com/6309070/59307579-a5012280-8c53-11e9-9e9e-67c2d25bacff.png)

Which is unexpected.

**Comment by smoothdeveloper** [comment]
Note there is similar problem currently with attributes on types:

![image](https://user-images.githubusercontent.com/87944/59648007-b8454e00-917d-11e9-9964-95d602970b7e.png)

here I'm switching a `module rec` to a smaller set of `and`ed types, and I have to move the `[<RequireQualifiedAccess>]` after the and.

I'm getting this before fixing it which is a bit misleading (attribute is valid on that language element):

![image](https://user-images.githubusercontent.com/87944/59648077-fc385300-917d-11e9-95db-e47fe794018d.png)

If it ends up being similar change for both and could be fixed, it would make the language a bit more lenient on those particular syntactic bits.

**Comment by Happypig375** [comment]
You mean #733?

**Comment by smoothdeveloper** [comment]
@Happypig375 correct!

Also interesting, this comment from @dsyme regarding such changes in parser: https://github.com/fsharp/fslang-suggestions/issues/514#issuecomment-332035672

**Comment by auduchinok** [comment]
Another case is indentation block for the binding/member:

<img width="347" alt="Screenshot 2019-07-04 at 13 45 30" src="https://user-images.githubusercontent.com/3923587/60661271-44d05b80-9e62-11e9-99d8-c83296ead06e.png">

**Comment by abelbraaksma** [comment]
Perhaps this and #514 can be merged, perhaps also #733 and #757, they basically all come down to the same thing:

* unifying the way the indentation rules apply to declarations-with-attributes, by allowing the attributes before, after and above the keyword.

**Comment by dsyme** [comment]
My understanding is this has been addressed as part of FS-1108, currently in preview https://github.com/fsharp/fslang-design/blob/main/preview/FS-1108-undentation-frenzy.md

**Comment by BentTranberg** [comment]
Ops, I'm late to the party. Again! On the same subject! Sorry if somebody is tired of me repeating this.

> This is already possible (or am I misunderstanding the suggestion?).
> 
> ```fsharp
> let [<Literal>] Foo = "foo"
> ```

That's what I have been recommending on several occasions in various places, because it avoids code creep. By placing an attribute inside the language construct to which it applies, rather than somewhere outside (e.g. before on the same line), you considerably lessen the risk of the language construct and the attribute drifting away from each other, with all the bad consequences that can have.

**Comment by Happypig375** [comment]
@dsyme I believe this should be closed as dupe of https://github.com/fsharp/fslang-suggestions/issues/514 rather than being considered completed, because this has not been fixed yet.

**Comment by dsyme** [comment]
> @dsyme I believe this should be closed as dupe of #514 rather than being considered completed, because this has not been fixed yet.

OK, thank you
